src/TFMiniPlus.cpp

```diff
@@ -11,72 +11,70 @@
  */
 
 #include "TFMiniPlus.h"
+#include "TFMiniPlusConstants.h"
 
-#define DATA_FRAME_MARKER 0x59
-#define DATA_FRAME_LENGTH 9
-#define CMD_FRAME_MARKER 0x5A
-#define MAX_CMD_RESPONSE_LENGTH 8
+template <std::size_t N>
+constexpr std::size_t buffer_size(const uint8_t (&buffer)[N]) noexcept {
+  return N;
+}
 
 void TFMiniPlus::begin(Stream* serial) { _stream = serial; }
 
 bool TFMiniPlus::readData() {
-  uint8_t buffer[9];
+  uint8_t buffer[DATA_FRAME_LENGTH];
 
-  TFMiniPlus::resetBuffer(_readDataBuffer, 9);
+  TFMiniPlus::resetBuffer(_readDataBuffer, buffer_size(_readDataBuffer));
 
   // skip to the first data frame header
   TFMiniPlus::skipToFrameHeader(DATA_FRAME_MARKER);
 
   // get the data body "all - frame marker byte"
-  _stream->readBytes(buffer, DATA_FRAME_LENGTH - 1);
+  _stream->readBytes(&buffer[1], DATA_FRAME_LENGTH - 1);
 
   // reconstruct the frame for checksum
-  uint8_t frame[DATA_FRAME_LENGTH] = {DATA_FRAME_MARKER};
-  for (uint8_t i = 1; i < DATA_FRAME_LENGTH; i++) {
-    frame[i] = buffer[i - 1];
-  }
+  buffer[0] = DATA_FRAME_MARKER;
+
   // validate data frame
-  if (TFMiniPlus::validateChecksum(frame, DATA_FRAME_LENGTH)) {
-    TFMiniPlus::copyBuffer(_readDataBuffer, frame, 9);
+  if (TFMiniPlus::validateChecksum(buffer, DATA_FRAME_LENGTH)) {
+    TFMiniPlus::copyBuffer(_readDataBuffer, buffer, buffer_size(_readDataBuffer), DATA_FRAME_LENGTH);
     return true;
   }
+
   return false;
 }
 
 uint16_t TFMiniPlus::getDistance() {
-  return uint16_t((_readDataBuffer[2] | (_readDataBuffer[3] << 8)));
```

Comment:
here, you were casting AFTER you made shifted the bits. This would lead to loss of data as the higher bytes are always truncated.
Example: High = 0xFF, Low = 0xAA
If you left bit shift the High byte in a uint8_t by 8, you get High = 0x00